lib/mayaUsd/undo/UsdUndoStateDelegate.cpp

```diff
@@ -0,0 +1,379 @@
+        .Msg("Inverting push field '%s' of '%s'\n", fieldName.GetText(), parentPath.GetText());
+
+    PopChild(parentPath, fieldName, value);
+}
+
+void UsdUndoStateDelegate::invertPopTokenChild(
+    const SdfPath& parentPath,
+    const TfToken& fieldName,
+    const TfToken& value)
+{
+    _setMessageAlreadyShowed = true;
+
+    TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+        .Msg("Inverting pop field '%s' of '%s'\n", fieldName.GetText(), parentPath.GetText());
+
+    PushChild(parentPath, fieldName, value);
+}
+
+void UsdUndoStateDelegate::invertPopPathChild(const SdfPath& parentPath, const TfToken& fieldName, const SdfPath& value)
+{
+    _setMessageAlreadyShowed = true;
+
+    TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+        .Msg("Inverting pop field '%s' of '%s'\n", fieldName.GetText(), parentPath.GetText());
+
+    PushChild(parentPath, fieldName, value);
+}
+
+void UsdUndoStateDelegate::_OnSetField(const SdfPath& path, const TfToken& fieldName, const VtValue& value)
+{
+    _MarkCurrentStateAsDirty();
+
+    // early return if we are not insdide an UsdUndoBlock
+    if (UsdUndoBlock::depth() == 0){
+        return;
+    }
+
+    if (!_setMessageAlreadyShowed) {
+        TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+            .Msg("Setting Field '%s' for Spec '%s'\n", fieldName.GetText(), path.GetText());
+    }
+
+    if (!_layer) {
+        return;
+    }
+
+    const VtValue inverseValue = _layer->GetField(path, fieldName);
+
+    UsdUndoManager::instance().addInverse(
+        std::bind(&UsdUndoStateDelegate::invertSetField, this, path, fieldName, inverseValue));
+}
+
+void UsdUndoStateDelegate::_OnSetField(
+    const SdfPath&                   path,
+    const TfToken&                   fieldName,
+    const SdfAbstractDataConstValue& value)
+{
+    _MarkCurrentStateAsDirty();
+
+    // early return if we are not insdide an UsdUndoBlock
+    if (UsdUndoBlock::depth() == 0){
+        return;
+    }
+
+    if (!_setMessageAlreadyShowed) {
+        TF_DEBUG(USDMAYA_UNDOSTATEDELEGATE)
+            .Msg("Setting Field '%s' for Spec '%s'\n", fieldName.GetText(), path.GetText());
+    }
+
+    if (!_layer) {
+        return;
+    }
+
+    const VtValue inverseValue = _layer->GetField(path, fieldName);
+
+    // add invert
+    UsdUndoManager::instance().addInverse(
+        std::bind(&UsdUndoStateDelegate::invertSetField, this, path, fieldName, inverseValue));
+}
+
+void UsdUndoStateDelegate::_OnSetFieldDictValueByKey(
+    const SdfPath& path,
+    const TfToken& fieldName,
+    const TfToken& keyPath,
+    const VtValue& value)
+{
+    throw std::logic_error(std::string { __func__ } + "is not implemented !!!");
+}
+
+void UsdUndoStateDelegate::_OnSetFieldDictValueByKey(
+    const SdfPath&                   path,
+    const TfToken&                   fieldName,
+    const TfToken&                   keyPath,
+    const SdfAbstractDataConstValue& value)
+{
+    throw std::logic_error(std::string { __func__ } + "is not implemented !!!");
```

Comment:
A would recommend using a TF_ERROR macro instead as USD is not a big fan of exceptions.

Reply:
@aloysbaillet Good point. Addressed in 98360899ab96ce4bf396784e32bd85c8bdcf78e0